analyzer/file/zip/zip.spicy

```diff
@@ -0,0 +1,205 @@
+
+module ZIP;
+
+import filter;
+import spicy;
+
+%byte-order = spicy::ByteOrder::Little;
+
+# We define only the compression levels that we support for content analysis.
+public type CompressionMethod = enum { Uncompressed = 0, Deflate = 8 };
+
+public type Archive = unit {
+    : Record[];
+};
+
+type Record = unit {
+    switch {
+        -> : File;
+        -> : DirectoryHeader;
+        -> : EndOfDirectory;
+        -> : EndOfDirectory64;
+        -> : DataDescriptor; # not sure if this can actually appear on its own
+    };
+};
+
+type DirectoryHeader = unit {
+    : b"PK\x01\x02";
+    version_made:           uint16;
+    version_needed:         uint16;
+    flags:                  bitfield(16) {
+        encrypted: 0 &convert=cast<bool>($$);
+        compression: 1..2;
+        has_descriptor: 3;
+        enhanced_deflation: 4;
+        compressed_patched: 5;
+        strong_encryption: 6;
+        efs: 11;
+        masked_headers: 13;
+    };
+
+    compression:            uint16 &convert=CompressionMethod($$);
+    time_:                  Time;
+    crc:                    uint32;
+    compressed_size:        uint32;
+    uncompressed_size:      uint32;
+    filename_length:        uint16;
+    extra_field_length:     uint16;
+    file_comment_length:    uint16;
+    disk_number_start:      uint16;
+    internal_attributes:    uint16;
+    external_attributes:    uint32;
+    offset_local_header:    uint32;
+    filename:               bytes &size=self.filename_length;
+    extra_field:            bytes &size=self.extra_field_length;
+    file_comment:           bytes &size=self.file_comment_length;
+};
+
+type EndOfDirectory = unit {
+    : b"PK\x05\x06";
+    disk_number:        uint16;
+    disk_cd_start:      uint16;
+    disk_cd_entries:    uint16;
+    cd_entries:         uint16;
+    cd_size:            uint32;
+    cd_offset:          uint32;
+    zip_comment_length: uint16;
+    zip_comment:        bytes &size=self.zip_comment_length;
+};
+
+type EndOfDirectory64 = unit {
+    : b"PK\x06\x06" { throw "ZIP64 not supported"; }
+};
+
+type DataDescriptor = unit {
+    :                   b"PK\x07\x08";
+    crc:                uint32;
+    compressed_size:    uint32;
+    uncompressed_size:  uint32;
+};
+
+type Time = unit {
+    raw: bitfield(32) {
+        sec:    0..4;
+        min:    5..10;
+        hour:   11..15;
+        day:    16..20;
+        month:  21..24;
+        year:   25..31;
+    };
+} &byte-order=spicy::ByteOrder::Little
+  &convert=spicy::mktime(self.raw.year + 1980, self.raw.month, self.raw.day, self.raw.hour, self.raw.min, self.raw.sec * 2);
+
+type File = unit {
+    : b"PK\x03\x04";
+    version:            uint16;
+    flags:              bitfield(16) {
+        encrypted: 0 &convert=cast<bool>($$);
+        compression: 1..2;
+        has_descriptor: 3;
+        enhanced_deflation: 4;
+        compressed_patched: 5;
+        strong_encryption: 6;
+        efs: 11;
+        masked_headers: 13;
+    };
```

Comment:
Since this seems identical tothe `flags` in `DirectoryHeader`, it would be nice to move the RHS into a field declaration, and reuse it here and in `DirectoryHeader`.

Reply:
Good point, will factor this out.